src/runtime/program/serialization.zig

```diff
@@ -0,0 +1,967 @@
+
+            // read and update Lamports
+            if (start + @sizeOf(u64) > memory.len) return InstructionError.InvalidArgument;
+            const lamports = std.mem.readInt(
+                u64,
+                memory[start .. start + @sizeOf(u64)][0..@sizeOf(u64)],
+                .little,
+            );
+            start += @sizeOf(u64);
+            if (borrowed_account.account.lamports != lamports) {
+                try borrowed_account.setLamports(lamports);
+            }
+
+            // add data length
+            start += @sizeOf(u64);
+            if (copy_account_data) {
+                if (start + pre_len > memory.len) return InstructionError.InvalidArgument;
+                const data = memory[start .. start + pre_len];
+                const can_data_be_resized =
+                    borrowed_account.checkCanSetDataLength(ic.tc.accounts_resize_delta, pre_len);
+                const can_data_be_mutated = borrowed_account.checkDataIsMutable();
+                if (can_data_be_resized == null and can_data_be_mutated == null) {
+                    try borrowed_account.setDataFromSlice(
+                        allocator,
+                        &ic.tc.accounts_resize_delta,
+                        data,
+                    );
+                } else {
+                    if (!std.mem.eql(u8, borrowed_account.account.data, data)) {
+                        if (can_data_be_resized) |err| return err;
+                        if (can_data_be_mutated) |err| return err;
+                    }
+                }
+            }
+        }
+
+        start += @sizeOf(Pubkey) // owner
+        + @sizeOf(u8) // executable
+        + @sizeOf(u64); // rent_epoch
+    }
+}
+
+fn deserializeParametersAligned(
+    allocator: std.mem.Allocator,
+    ic: *InstructionContext,
+    copy_account_data: bool,
+    memory: []u8,
+    account_lengths: []const usize,
+) (error{OutOfMemory} || InstructionError)!void {
+    var start: usize = @sizeOf(u64);
+
+    for (0..ic.info.account_metas.len) |index_in_instruction_| {
+        const index_in_instruction: u16 = @intCast(index_in_instruction_);
+        const account_meta = ic.info.account_metas.buffer[index_in_instruction];
+        const pre_len = account_lengths[index_in_instruction];
+
+        start += @sizeOf(u8);
+        if (account_meta.index_in_callee != index_in_instruction) {
+            start += 7;
+        } else {
+            var borrowed_account = try ic.borrowInstructionAccount(index_in_instruction);
+            defer borrowed_account.release();
+
+            start += @sizeOf(u8) // is_signer
+            + @sizeOf(u8) // is_writable
+            + @sizeOf(u8) // executable
+            + @sizeOf(u32) // original data len
+            + @sizeOf(Pubkey); // key
+
+            // read owner
+            if (start + @sizeOf(Pubkey) > memory.len) return InstructionError.InvalidArgument;
+            const owner = memory[start .. start + @sizeOf(Pubkey)];
```

Comment:
```suggestion
            const owner = memory[start ..][0..@sizeOf(Pubkey)];
```
Ditto for the other usages.

Reply:
bde9467a